File: frontend/src/api/rating.ts
import { api } from './client'
import type { RatingEntry, RatingPeriod, StatsLockReason } from '../types'

interface RawRatingItem {
  rank: number
  runner_id: number
  first_name: string
  last_name: string
  avatar?: string | null
  finished_count: number
}

interface RawRatingResponse {
  period: string
  entries: RawRatingItem[]
  total: number
  page: number
  page_size: number
  me: RawRatingItem | null
  lock_reason: StatsLockReason
  missing_fields: string[]
}

function mapItem(e: RawRatingItem): RatingEntry {
  return {
    rank: e.rank,
    user_id: e.runner_id,
    first_name: e.first_name,
    last_name: e.last_name,
    avatar_url: e.avatar ?? null,
    city: null,
    finished_count: e.finished_count,
    score: e.finished_count,
  }
}

export const ratingApi = {
  list: async (
    period: RatingPeriod = 'all',
    page = 1,
  ): Promise<{
    entries: RatingEntry[]
    total: number
    page: number
    pageSize: number
    me: RatingEntry | null
    lockReason: StatsLockReason
    missingFields: string[]
  }> => {
    // No `auth: false` here — the request client already attaches the token
    // when a session exists, and the endpoint stays public either way; being
    // logged in just adds a personalized `me` row when off the current page.
    const res = await api.get<RawRatingResponse>('/rating', { query: { period, page } })
    return {
      entries: res.entries.map(mapItem),
      total: res.total,
      page: res.page,
      pageSize: res.page_size,
      me: res.me ? mapItem(res.me) : null,
      lockReason: res.lock_reason,
      missingFields: res.missing_fields,
    }
  },
}

File: backend/app/api/rating.py
from typing import Literal

from fastapi import APIRouter, Query

from app.api.deps import OptionalUser, SessionDep
from app.schemas.rating import RatingItem, RatingResponse
from app.services.profile_completeness_service import stats_access_lock
from app.services.rating_service import compute_rating

router = APIRouter(tags=["rating"])

PAGE_SIZE = 20


@router.get("/rating", response_model=RatingResponse)
async def rating(
    session: SessionDep,
    user: OptionalUser,
    period: Literal["all", "year", "month"] = "all",
    page: int = Query(1, ge=1),
) -> RatingResponse:
    lock_reason, missing_fields = await stats_access_lock(session, user)
    if lock_reason is not None:
        return RatingResponse(
            period=period,
            entries=[],
            total=0,
            page=1,
            page_size=PAGE_SIZE,
            lock_reason=lock_reason,
            missing_fields=missing_fields,
        )

    # compute_rating already drops anyone with 0 activity in the window, so the
    # ranking never lists someone sitting at 0 for the selected period.
    entries = await compute_rating(session, period)
    items = [
        RatingItem(
            rank=i,
            runner_id=e.runner_id,
            first_name=e.first_name,
            last_name=e.last_name,
            avatar=e.avatar,
            finished_count=e.finished_count,
        )
        for i, e in enumerate(entries, start=1)
    ]

    total = len(items)
    total_pages = max(1, (total + PAGE_SIZE - 1) // PAGE_SIZE)
    page = min(page, total_pages)
    start = (page - 1) * PAGE_SIZE
    page_items = items[start : start + PAGE_SIZE]

    # Only surface a personal "me" row when the user is ranked but *not* on the
    # page being viewed — otherwise they're already visible. Its rank tells the
    # frontend which page to highlight.
    me = None
    if user is not None:
        on_page = {it.runner_id for it in page_items}
        me = next(
            (it for it in items if it.runner_id == user.id and it.runner_id not in on_page),
            None,
        )

    return RatingResponse(
        period=period,
        entries=page_items,
        total=total,
        page=page,
        page_size=PAGE_SIZE,
        me=me,
    )
